fuel-vm/src/interpreter/blockchain.rs

```diff
@@ -257,161 +568,166 @@ where
-        let key = unsafe { Bytes32::as_ref_unchecked(&self.memory[a..ax]) };
+    // Safety: Memory bounds logically verified by the interpreter
+    let contract = unsafe { ContractId::as_ref_unchecked(&memory[d..dx]) };
+    let key = unsafe { Bytes32::as_ref_unchecked(&memory[a..ax]) };
 
-        let mut value = Bytes32::default();
+    let mut value = Bytes32::default();
 
-        value[..WORD_SIZE].copy_from_slice(&c.to_be_bytes());
-
-        let result = self
-            .storage
-            .merkle_contract_state_insert(contract, key, &value)
-            .map_err(RuntimeError::from_io)?;
-
-        self.registers[rb] = result.is_some() as Word;
-
-        self.inc_pc()
-    }
-
-    pub(crate) fn state_write_qword(&mut self, a: Word, rb: RegisterId, c: Word, d: Word) -> Result<(), RuntimeError> {
-        Self::is_register_writable(rb)?;
-        let contract_id = *self.internal_contract()?;
-        let input = StateWriteQWord::new(a, c, d)?;
-        let Self {
-            ref mut storage,
-            ref mut memory,
-            ref mut registers,
-            ..
-        } = self;
+    value[..WORD_SIZE].copy_from_slice(&c.to_be_bytes());
 
-        state_write_qword(&contract_id, storage, memory.as_mut(), &mut registers[rb], input)?;
-        self.inc_pc()
-    }
+    let result = storage
+        .merkle_contract_state_insert(contract, key, &value)
+        .map_err(RuntimeError::from_io)?;
 
-    pub(crate) fn timestamp(&mut self, ra: RegisterId, b: Word) -> Result<(), RuntimeError> {
-        Self::is_register_writable(ra)?;
+    *exists = result.is_some() as Word;
 
-        self.block_height()
-            .and_then(|c| (b <= c as Word).then_some(()).ok_or(PanicReason::TransactionValidity))?;
+    inc_pc(pc)
+}
 
-        let b = u32::try_from(b).map_err(|_| PanicReason::ArithmeticOverflow)?;
+pub(crate) fn timestamp(
+    storage: &impl InterpreterStorage,
+    block_height: u32,
+    pc: RegMut<PC>,
+    result: &mut Word,
+    b: Word,
+) -> Result<(), RuntimeError> {
+    (b <= block_height as Word)
+        .then_some(())
+        .ok_or(PanicReason::TransactionValidity)?;
 
-        self.registers[ra] = self.storage.timestamp(b).map_err(|e| e.into())?;
+    let b = u32::try_from(b).map_err(|_| PanicReason::ArithmeticOverflow)?;
 
-        self.inc_pc()
-    }
+    *result = storage.timestamp(b).map_err(|e| e.into())?;
 
-    pub(crate) fn message_output(&mut self, a: Word, b: Word, c: Word, d: Word) -> Result<(), RuntimeError> {
-        let ax = checked_add_word(a, Address::LEN as Word)?;
-        let bx = checked_add_word(ax, b)?;
+    inc_pc(pc)
+}
 
-        //TODO check on b vs MEM_MAX_ACCESS_SIZE is looser than msg length check
-        if b > self.params.max_message_data_length || b > MEM_MAX_ACCESS_SIZE || bx > VM_MAX_RAM {
-            return Err(PanicReason::MemoryOverflow.into());
-        }
+struct MessageOutputInput<'vm, Tx> {
```

Comment:
I find it somewhat counter-intuitive to have "Input" objects that also do the real execution. As I wouldn't expect an input struct to perform complicated tasks. Maybe something like `Operation` instead of `Input` would be easier to follow?

```suggestion
struct MessageOutputOperation<'vm, Tx> {
```

Reply:
Another common pattern is to use `Context`/`Ctx` when providing the context for some operation, e.g.
```suggestion
struct MessageOutputCtx<'vm, Tx> {
```